### PR TITLE
[ARM] Don't use -1 as invalid register number in assembly parser.

### PR DESCRIPTION
**Pull request by topperc**
Use MCRegister instead.

### CODE REVIEW

File: llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp [hunk 1]
```diff
@@ -4634,7 +4633,7 @@ bool ARMAsmParser::parseRegisterList(OperandVector &Operands, bool EnforceOrder,
     Reg = getDRegFromQReg(Reg);
     EReg = MRI->getEncodingValue(Reg);
     Registers.emplace_back(EReg, Reg);
-    ++Reg;
+    Reg = Reg + 1;
```
Comment (topperc) [inline llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:4636]: Should we add a operator++ to MCRegister?

Comment (s-barannikov) [reply llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:4636]: IIRC registers are emitted in the order they are defined in td, which makes ++ work most of the time, but there is no guarantee in general.

In this particular case it should probably be `Reg = MRI->getSubReg(Reg, ARM::dsub_1)` (see `getDRegFromQReg`).

Comment (s-barannikov) [reply llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:4636]: > IIRC registers are emitted in the order they are defined in td

No, it seems it's not that simple. They're not sorted alphabetically either.

**Review by s-barannikov — APPROVED** [review]
LGTM

**Review by AlfieRichardsArm — APPROVED** [review]
LGTM! Seems like a really nice improvement to legability

**Pull request merged**